src/main/java/co/elastic/logstash/filters/elasticintegration/geoip/IpDatabaseAdapter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V.
+ * under one or more contributor license agreements. Licensed under the
+ * Elastic License 2.0; you may not use this file except in compliance
+ * with the Elastic License 2.0.
+ */
+package co.elastic.logstash.filters.elasticintegration.geoip;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.common.CheckedBiFunction;
+import org.elasticsearch.ingest.geoip.IpDatabase;
+import org.elasticsearch.ingest.geoip.shaded.com.maxmind.db.CHMCache;
+import org.elasticsearch.ingest.geoip.shaded.com.maxmind.db.NoCache;
+import org.elasticsearch.ingest.geoip.shaded.com.maxmind.db.NodeCache;
+import org.elasticsearch.ingest.geoip.shaded.com.maxmind.db.Reader;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Optional;
+
+public class IpDatabaseAdapter implements IpDatabase {
+    private static final Logger LOGGER = LogManager.getLogger(IpDatabaseAdapter.class);
+
+    private final Reader databaseReader;
+    private final String databaseType;
+
+    public IpDatabaseAdapter(final Reader databaseReader) {
+        this.databaseReader = databaseReader;
+        this.databaseType = databaseReader.getMetadata().getDatabaseType();
+    }
+
+    @Override
+    public String getDatabaseType() {
+        return this.databaseType;
+    }
+
+    @Override
+    public <RESPONSE> RESPONSE getResponse(String ipAddress, CheckedBiFunction<Reader, String, RESPONSE, Exception> responseProvider) {
+        try {
+            return responseProvider.apply(this.databaseReader, ipAddress);
+        } catch (Exception e) {
+            throw ExceptionsHelper.convertToRuntime(e);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        // no-op
+        // IpDatabase is an AutoCloseable class which is not our intention to close after try-with-resource operations.
+        // use closeReader() instead
+    }
+
+    public void closeReader() throws IOException {
+        LOGGER.debug("Closing the database adapter");
+        this.databaseReader.close();
+    }
```

Comment:
review note: Addition to spike diff, introduced `IpDatabaseAdapter#closeReader()` method to intentionally close the database reader. The `IpDatabase` is an `AutoCloseable` class which the intention is automatically close database with try-with-resource (`try (IpDatabaseAdapter ipAdapter = new IpDatabaseAdapter())`) operations in Java 7. We have Java 6 style source, the auto close is being called when we try safely load DB in `ManagedIpDatabaseHolder#loadDatabase() -> try {...} catch {..}`.